Include/internal/pycore_object.h

```diff
@@ -25,6 +27,18 @@ extern "C" {
     }
 
 
+static inline bool _PyObject_IsSingleton(PyObject *obj)
+{
+#define SINGLETONS (&_PyRuntime.global_objects.singletons)
+    if (((uintptr_t)obj >= (uintptr_t)(SINGLETONS)) &&
```

Comment:
I think this may still be arguably unspecified behavior, though probably not an issue on most compilers.

https://stackoverflow.com/questions/57639853/is-comparing-two-pointers-with-undefined-behavior-if-they-are-both-cast-to-an

Reply:
I've fixed it.  Casting each with `(uint64_t)(void *)` should be sufficient to play it safe.

FTR, I looked into the [C99 spec](https://port70.net/~nsz/c/c99/n1256.html), particularly 6.5.8 and 7.18.1.4.  Also note that the builtin `id()` relies on the uniqueness of the values resulting from the `(uint64_t)(void *)` cast.

Reply:
I'm still not 100% convinced that this is within the spec. In particular, I don't think there's anything in the spec that prevents `(uintptr_t)ptr` from rotating the bits or reversing the bits or inverting all bits or something strange like that. The spec guarantees that `ptr == (void *)(uintptr_t)ptr`, so the mapping `integers corresponding to pointers --> pointers` respects equality/unequality, but it doesn't specify that the mapping respects `<=`. I'm not sure you can prove the statement "If `uintptr_t i, j` satisfy `i <= j` then `(void *)i <= (void *)j`" using only the assumptions of the standard. Comparing `id(x) != id(y)` is equivalent to `x is not y`, but I'm not sure `id(x) <= id(y)` guarantees that `pyobject_ptr_x <= pyobject_ptr_y`, unless the objects live in the same buffer.

Again, this is probably not a real problem on any relevant compilers, so it's probably okay to leave the PR as is.

Reply:
[Footnote 67](http://port70.net/~nsz/c/c11/n1570.html#note67) says """The mapping functions for converting a pointer to an integer or an integer to a pointer are intended to be consistent with the addressing structure of the execution environment""" so probably nothing *too* obscure is "intended" to happen.

Reply:
Thanks for all the insight.

Reply:
Just to make sure that the `pointer <--> integer` mapping isn't "inverted", there could maybe be an assertion like

```c
assert((uintptr_t)(SINGLETONS) < (uintptr_t)((SINGLETONS) + 1));
```